### PR TITLE
Alias traits from druntime

### PR DESCRIPTION
**Pull request by TurkeyMan**
This is complement to https://github.com/dlang/druntime/pull/2742

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Ping

**Comment by RazvanN7** [comment]
Ping @TurkeyMan

**Review by RazvanN7 — APPROVED** [review]

File: std/traits.d [hunk 1]
```diff
@@ -2718,7 +2714,7 @@ alias FieldTypeTuple = Fields;
     static assert(is(FieldTypeTuple!NestedClass == AliasSeq!int));
 
     static interface I {}
-    static assert(is(Fields!I == AliasSeq!()));
+    static assert(is(Fields!I == AliasSeq!(I)));
```
Comment (MoonlightSentinel) [inline std/traits.d:2717]: > Fix unittest

This is a regression... The Druntime implementation was not updated alongside phobos

Comment (RazvanN7) [reply std/traits.d:2717]: Fixes: https://github.com/dlang/druntime/pull/3569 ; https://github.com/dlang/phobos/pull/8241 .

Will re-add the test to phobos once druntime is fixed.

**Comment by nordlow** [comment]
> This is complement to [dlang/druntime#2742](https://github.com/dlang/druntime/pull/2742)

The template XXXTypeOf are anti-patterns and should be deprecated from phobos. It's unfortunate that we now also contaminate druntime with them. I believe @andralex agrees with me.

File: std/traits.d (hunk 1 above)
Comment (andralex) [reply std/traits.d:2717]: This is yet another piece of evidence that code duplication is bad.

File: std/traits.d [hunk 2]
```diff
@@ -6025,17 +6014,8 @@ template StaticArrayTypeOf(T)
...
-        alias X = OriginalType!T;
-
-    static if (is(X == E[], E))
-    {
-        alias DynamicArrayTypeOf = X;
-    }
-    else
-        static assert(0, T.stringof~" is not a dynamic array");
+    import core.internal.traits : _DynamicArrayTypeOf = DynamicArrayTypeOf;
+    alias DynamicArrayTypeOf = _DynamicArrayTypeOf!T;
```
Comment (andralex) [inline std/traits.d:6018]: Instead of giving it a name, why not import it statically?

**Comment by andralex** [comment]
What was the usefulness of those `XXXTypeOf` things?

**Comment by nordlow** [comment]
Moreover, FYI, there has recently been an effort to make phobos traits wrapping druntime traits become symbol aliases instead of real template to avoid extra template instances. Now we are creating new wrappers instead of them being symbolic aliases.

**Comment by nordlow** [comment]
> What was the usefulness of those `XXXTypeOf` things?

Basically, for instance, `BooleanTypeOf!(T)` returns a bool if T is either a 

- a bool or
- a struct that alias thises a bool.

**Comment by andralex** [comment]
oh god yes those must go

File: std/traits.d (hunk 2 above)
Comment (nordlow) [reply std/traits.d:6018]: Can you show a code example of what you mean by import statically?

Comment (thewilsonator) [reply std/traits.d:6018]: I presume he means `import core.internal.traits :  DynamicArrayTypeOf;`

**Comment by RazvanN7** [comment]
>  Now we are creating new wrappers instead of them being symbolic aliases.

This PR just got rid of the duplication. We are now free to get rid of the wrappers; that is if we do not decide to get rid of this functionality altogether.

**Comment by RazvanN7** [comment]
> Have you made any progress on ddoc improvements, @RazvanN7?

Nope.

**Comment by nordlow** [comment]
> > Have you made any progress on ddoc improvements, @RazvanN7?
> Nope.

What kind of improvement are you considering?

**Comment by nordlow** [comment]
> > Now we are creating new wrappers instead of them being symbolic aliases.
> 
> This PR just got rid of the duplication. We are now free to get rid of the wrappers; that is if we do not decide to get rid of this functionality altogether.

Would you approve me apply the ddoc-trick to these aswell?

**Pull request merged**